pkgs/_analyzer_macros/lib/analyzer_macro_implementation.dart

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) 2024, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:_macro_host/macro_host.dart';
+// ignore: implementation_imports
+import 'package:analyzer/src/summary2/macro_injected_impl.dart' as injected;
+import 'package:dart_model/dart_model.dart';
+import 'package:macro_service/macro_service.dart';
+import 'package:macros/macros.dart';
+// ignore: implementation_imports
+import 'package:macros/src/executor.dart' as injected;
+
+/// Injected macro implementation for the analyzer.
+class MacroImplementation implements injected.MacroImplementation, HostService {
+  final Uri packageConfig;
+  final Map<String, String> macroImplByName;
+  late final MacroHost _host;
+
+  MacroImplementation._(this.packageConfig, this.macroImplByName);
+
+  /// Starts the macro host.
+  ///
+  /// [packageConfig] is the package config of the workspace of the code being
+  /// edited.
+  ///
+  /// [macroImplByName] identifies macros, it's a placeholder until we identify
+  /// macros using package config. Keys are macro annotation qualified names
+  /// (`uri#name`) and values are macro implementation qualified names.
+  static Future<MacroImplementation> start({
+    required Uri packageConfig,
+    required Map<String, String> macroImplByName,
+  }) async {
+    final result = MacroImplementation._(packageConfig, macroImplByName);
+    result._host = await MacroHost.serve(
+        macroImplByName: macroImplByName, service: result);
+    return result;
+  }
+
+  @override
+  injected.MacroPackageConfigs get packageConfigs => MacroPackageConfigs(this);
+
+  @override
+  injected.MacroRunner get runner => MacroRunner(this);
+
+  @override
+  Future<Response?> handle(MacroRequest request) async {
+    if (request.type != MacroRequestType.queryRequest) return null;
+    return Response.queryResponse(QueryResponse());
+  }
+}
+
+class MacroPackageConfigs implements injected.MacroPackageConfigs {
+  final MacroImplementation impl;
+
+  MacroPackageConfigs(this.impl);
+
+  @override
+  bool isMacro(Uri uri, String name) =>
+      impl._host.isMacro(impl.packageConfig, QualifiedName('$uri#$name'));
+}
+
+class MacroRunner implements injected.MacroRunner {
+  final MacroImplementation impl;
+
+  MacroRunner(this.impl);
+
+  @override
+  injected.RunningMacro run(Uri uri, String name) => RunningMacro.run(
+      impl,
+      QualifiedName('$uri#$name'),
+      // Look up from the macro name to its implementation.
+      QualifiedName(impl.macroImplByName['$uri#$name']!));
+}
+
+class RunningMacro implements injected.RunningMacro {
+  final MacroImplementation impl;
+  final QualifiedName name;
+  final QualifiedName implementation;
+  late final Future _started;
+
+  RunningMacro._(this.impl, this.name, this.implementation);
+
+  static RunningMacro run(MacroImplementation impl, QualifiedName name,
```

Comment:
Can this instead return a `Future<RunningMacro>` and we can drop the `_started` field and the awaits below?

Reply:
Yeeeah that's what I had originally, but the `RunningMacro` method that gets injected in the analyzer is sync so this has to be sync. I added a TODO to consider changing that upstream :)